prusti-viper/src/encoder/name_interner.rs

```diff
@@ -0,0 +1,129 @@
+// © 2021, ETH Zurich
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+use std::collections::{HashMap, HashSet};
+use log::debug;
+
+/// Name interner.
+/// This structure can be used to shorten long unique names without losing the uniqueness property.
+#[derive(Debug)]
+pub struct NameInterner {
+    name_to_symbol: HashMap<String, String>,
+    used_symbols: HashSet<String>,
+}
+
+impl NameInterner {
+    pub fn new() -> Self {
+        NameInterner {
+            name_to_symbol: HashMap::new(),
+            used_symbols: HashSet::new(),
+        }
+    }
+
+    /// Intern a full unique name, returning a possibly readable string that uniquely identifies it.
+    /// The `readable_names` must not collide with past or future `full_unique_name`s.
+    pub fn intern<S: ToString>(&mut self, full_unique_name: S, readable_names: &[S]) -> String {
+        let full_unique_name = full_unique_name.to_string();
+        let readable_names: Vec<_> = readable_names.iter().map(|s| s.to_string()).collect();
+        assert!(!readable_names.contains(&"".to_string()));
```

Comment:
`readable_names` is not used until after the `if let` check; it can be calculated futher down. The assertion is a debug assertion?

Reply:
Regarding debug, as above. Regarding the position, I think it's best to check them as early as possible. So, I can move one of the following assertions; the other needs to stay after the `if let`.

Reply:
Actually, both the last two assertions need to stay after the `if let`.

Reply:
In the end I'm using `AsRef<str>` instead of `ToString`, to avoid the allocation at all.